src/expr/src/explain.rs

```diff
@@ -183,8 +183,16 @@ impl<'a> ViewExplanation<'a> {
     /// Attach type information into the explanation.
     pub fn explain_types(&mut self) {
         for node in &mut self.nodes {
-            // TODO(jamii) `typ` is itself recursive, so this is quadratic :(
-            node.typ = Some(node.expr.typ());
+            if let MirRelationExpr::Let { .. } = &node.expr {
+                // Skip.
+                // Since we don't print out Let nodes in the explanation,
+                // types of Let nodes should not be attached to the
+                // explanation. The type information of a Let is always the
+                // same as the the type of the body.
+            } else {
+                // TODO(jamii) `typ` is itself recursive, so this is quadratic :(
```

Comment:
Unrelated to this patch, but I've just realized that `nodes` is in post-order which means that we could compute the types incrementally using `typ_with_input_types`, so we can resolve this TODO in a separate PR.

Reply:
I realized the same thing. I had originally intended to add computing types incrementally using `typ_with_input_types` to this PR, but my first attempt caused Materialize to crash when explaining plans with multiple let statements, so I figured incremental type computation was complicated enough to merit a separate PR.

Reply:
Created follow up issue #9030 to track this.